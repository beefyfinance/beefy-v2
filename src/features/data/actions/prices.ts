import { createAsyncThunk } from '@reduxjs/toolkit';
import { BeefyAPITokenPricesResponse } from '../apis/beefy';
import { getBeefyApi } from '../apis/instances';

export const fetchAllPricesAction = createAsyncThunk<BeefyAPITokenPricesResponse, {}>(
  'prices/fetchAllPricesAction',
  async () => {
    const api = getBeefyApi();
    const [prices, lpPrices] = await Promise.all([api.getPrices(), api.getLPs()]);
    return {
      ...prices,
      ...lpPrices,
    };
  }
);
