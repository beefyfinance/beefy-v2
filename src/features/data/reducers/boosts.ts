import { createSlice } from '@reduxjs/toolkit';
import { WritableDraft } from 'immer/dist/internal';
import { fetchAllBoosts } from '../actions/boosts';
import { BoostConfig } from '../apis/config';
import { BoostEntity, isBoostActive, isBoostPreStake } from '../entities/boost';
import { ChainEntity } from '../entities/chain';
import { VaultEntity } from '../entities/vault';
import { NormalizedEntity } from '../utils/normalized-entity';

/**
 * State containing Vault infos
 */
export type BoostsState = NormalizedEntity<BoostEntity> & {
  byVaultId: {
    [vaultId: VaultEntity['id']]: {
      allBoostsIds: BoostEntity['id'][];
      prestakeBoostsIds: BoostEntity['id'][];
      activeBoostsIds: BoostEntity['id'][];
    };
  };
  byChainId: {
    [chainId: ChainEntity['id']]: {
      allBoostsIds: BoostEntity['id'][];
      prestakeBoostsIds: BoostEntity['id'][];
      activeBoostsIds: BoostEntity['id'][];
    };
  };
};
export const initialBoostsState: BoostsState = {
  byId: {},
  allIds: [],
  byVaultId: {},
  byChainId: {},
};

export const boostsSlice = createSlice({
  name: 'boosts',
  initialState: initialBoostsState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    // when boost list is fetched, add all new tokens
    builder.addCase(fetchAllBoosts.fulfilled, (sliceState, action) => {
      for (const [chainId, boosts] of Object.entries(action.payload)) {
        for (const boost of boosts) {
          addBoostToState(sliceState, chainId, boost);
        }
      }
    });
  },
});

function addBoostToState(
  sliceState: WritableDraft<BoostsState>,
  chainId: ChainEntity['id'],
  apiBoost: BoostConfig
) {
  if (apiBoost.id in sliceState.byId) {
    return;
  }

  const boost: BoostEntity = {
    id: apiBoost.id,
    chainId: chainId,
    assets: apiBoost.assets,
    earnedTokenId: apiBoost.earnedOracleId,
    earnContractAddress: apiBoost.earnContractAddress,
    logo: apiBoost.logo,
    name: apiBoost.name,
    partnerIds: apiBoost.partners ? apiBoost.partners.map(p => p.website) : [],
    status: apiBoost.status as BoostEntity['status'],
    vaultId: apiBoost.poolId,
  };
  sliceState.byId[boost.id] = boost;
  sliceState.allIds.push(boost.id);

  // add to vault id index
  if (sliceState.byVaultId[boost.vaultId] === undefined) {
    sliceState.byVaultId[boost.vaultId] = {
      allBoostsIds: [],
      activeBoostsIds: [],
      prestakeBoostsIds: [],
    };
  }
  sliceState.byVaultId[boost.vaultId].allBoostsIds.push(boost.id);
  if (isBoostActive(boost)) {
    sliceState.byVaultId[boost.vaultId].activeBoostsIds.push(boost.id);
  }
  if (isBoostPreStake(boost)) {
    sliceState.byVaultId[boost.vaultId].prestakeBoostsIds.push(boost.id);
  }

  // add to chain id index
  if (sliceState.byChainId[chainId] === undefined) {
    sliceState.byChainId[chainId] = {
      allBoostsIds: [],
      activeBoostsIds: [],
      prestakeBoostsIds: [],
    };
  }
  sliceState.byChainId[chainId].allBoostsIds.push(boost.id);
  if (isBoostActive(boost)) {
    sliceState.byChainId[chainId].activeBoostsIds.push(boost.id);
  }
  if (isBoostPreStake(boost)) {
    sliceState.byChainId[chainId].prestakeBoostsIds.push(boost.id);
  }
}
