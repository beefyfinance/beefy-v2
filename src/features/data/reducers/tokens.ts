import { createSlice } from '@reduxjs/toolkit';
import BigNumber from 'bignumber.js';
import { WritableDraft } from 'immer/dist/internal';
import { BeefyState } from '../../../redux-types';
import { fetchAllBoosts } from '../actions/boosts';
import { fetchChainConfigs } from '../actions/chains';
import { fetchAllPricesAction } from '../actions/prices';
import {
  fetchAddressBookAction,
  FetchAddressBookPayload,
  fetchAllAddressBookAction,
} from '../actions/tokens';
import { fetchAllVaults } from '../actions/vaults';
import { BoostConfig, VaultConfig } from '../apis/config';
import { ChainEntity } from '../entities/chain';
import {
  isTokenErc20,
  isTokenNative,
  TokenEntity,
  TokenErc20,
  TokenNative,
} from '../entities/token';
import { selectChainById } from '../selectors/chains';
import {
  getBoostTokenIdFromLegacyConfig,
  getOracleTokenFromLegacyVaultConfig,
} from '../utils/config-hacks';

/**
 * State containing Vault infos
 */
export type TokensState = {
  // we need to split by chain because tokens from different chains have the same ids
  byChainId: {
    [chainId: ChainEntity['id']]: {
      byId: {
        [id: string]: TokenEntity;
      };
      native: TokenNative['id'];
      wnative: TokenErc20['id'];
      /**
       * we keep the list of tokens where we could be interested in fetching the balance of
       * it would be more correct to put those inside the balance reducer but this token
       * reducer has a number of config fixes that I find would make for a more complex code
       * if refactored. And we have to update the config anyway to make it smaller, so move this
       * inside the balance reducer once the config is reworked
       */
      interestingBalanceTokenIds: TokenEntity['id'][];
    };
  };
  prices: {
    byTokenId: {
      [tokenId: TokenEntity['id']]: BigNumber;
    };
  };
};
export const initialTokensState: TokensState = {
  byChainId: {},
  prices: { byTokenId: {} },
};

export const tokensSlice = createSlice({
  name: 'tokens',
  initialState: initialTokensState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    // handle native token config
    builder.addCase(fetchChainConfigs.fulfilled, (sliceState, action) => {
      for (const chainConf of action.payload.chainConfigs) {
        const chainId = chainConf.id;

        if (sliceState.byChainId[chainId] === undefined) {
          sliceState.byChainId[chainId] = {
            byId: {},
            interestingBalanceTokenIds: [],
            native: null,
            wnative: null,
          };
        }

        const tokenId = chainConf.walletSettings.nativeCurrency.symbol;
        const token: TokenNative = {
          id: tokenId,
          chainId: chainId,
          decimals: 18, // TODO: not sure about that
          address: null,
          symbol: chainConf.walletSettings.nativeCurrency.symbol,
          type: 'native',
          buyUrl: sliceState.byChainId[chainId].byId[tokenId]?.buyUrl ?? null,
          website: sliceState.byChainId[chainId].byId[tokenId]?.website ?? null,
          description: sliceState.byChainId[chainId].byId[tokenId]?.description ?? null,
        };
        sliceState.byChainId[chainId].byId[token.id] = token;
        sliceState.byChainId[chainId].native = token.id;
        sliceState.byChainId[chainId].interestingBalanceTokenIds.push(token.id);
      }
    });

    // when vault list is fetched, add all new tokens
    builder.addCase(fetchAllVaults.fulfilled, (sliceState, action) => {
      for (const [chainId, vaults] of Object.entries(action.payload.byChainId)) {
        const chain = selectChainById(action.payload.state, chainId);
        for (const vault of vaults) {
          addVaultToState(action.payload.state, sliceState, chain, vault);
        }
      }
    });

    // when boost list is fetched, add all new tokens
    builder.addCase(fetchAllBoosts.fulfilled, (sliceState, action) => {
      for (const [chainId, boosts] of Object.entries(action.payload)) {
        for (const boost of boosts) {
          addBoostToState(sliceState, chainId, boost);
        }
      }
    });

    // when prices are changed, update prices
    // this could also just be a a super quick drop in replacement
    // if we are OK to not use BigNumber, which I don't think we are
    builder.addCase(fetchAllPricesAction.fulfilled, (sliceState, action) => {
      for (const tokenId of Object.keys(action.payload)) {
        let tokenPrice = action.payload[tokenId];

        // when the api fails to fetch the token price, we say 1 token = $1
        if (tokenPrice === null || tokenPrice === undefined) {
          console.warn(`API returned an empty price for token ${tokenId}`);
          tokenPrice = 1.0;
        }

        // new price, add it
        if (sliceState.prices.byTokenId[tokenId] === undefined) {
          sliceState.prices.byTokenId[tokenId] = new BigNumber(tokenPrice);

          // price exists, update it if it changed
        } else if (sliceState.prices.byTokenId[tokenId].comparedTo(tokenPrice) === 0) {
          sliceState.prices.byTokenId[tokenId] = new BigNumber(tokenPrice);
        }
      }
    });

    // we have another way of finding token info
    builder.addCase(fetchAddressBookAction.fulfilled, (sliceState, action) => {
      addAddressBookToState(sliceState, action.payload);
    });
    builder.addCase(fetchAllAddressBookAction.fulfilled, (sliceState, action) => {
      for (const payload of action.payload) {
        addAddressBookToState(sliceState, payload);
      }
    });
  },
});

function addAddressBookToState(
  sliceState: WritableDraft<TokensState>,
  addressBookPayload: FetchAddressBookPayload
) {
  const chainId = addressBookPayload.chainId;

  if (sliceState.byChainId[chainId] === undefined) {
    sliceState.byChainId[chainId] = {
      byId: {},
      interestingBalanceTokenIds: [],
      native: null,
      wnative: null,
    };
  }

  for (const [addressBookId, token] of Object.entries(addressBookPayload.addressBook)) {
    if (sliceState.byChainId[chainId].byId[token.id] === undefined) {
      sliceState.byChainId[chainId].byId[token.id] = token;
    } else {
      const existingToken = sliceState.byChainId[chainId].byId[token.id];
      existingToken.buyUrl = existingToken.buyUrl || token.buyUrl;
      existingToken.description = existingToken.description || token.description;
      existingToken.website = existingToken.website || token.website;
    }
    if (addressBookId === 'WNATIVE' && !sliceState.byChainId[chainId].wnative) {
      sliceState.byChainId[chainId].wnative = token.id;
    }

    // update the native token address for those chains who have one
    if (isTokenNative(token) && token.address !== null) {
      const chainNative = sliceState.byChainId[chainId].byId[token.id] as TokenNative;
      if (chainNative && chainNative.address === null) {
        chainNative.address = token.address;
      }
    }
  }
}

function addBoostToState(
  sliceState: WritableDraft<TokensState>,
  chainId: ChainEntity['id'],
  apiBoost: BoostConfig
) {
  if (sliceState.byChainId[chainId] === undefined) {
    sliceState.byChainId[chainId] = {
      byId: {},
      interestingBalanceTokenIds: [],
      native: null,
      wnative: null,
    };
  }

  const tokenId = getBoostTokenIdFromLegacyConfig(apiBoost);
  if (sliceState.byChainId[chainId].byId[tokenId] === undefined) {
    const token: TokenEntity = {
      id: tokenId,
      chainId: chainId,
      contractAddress: apiBoost.earnedTokenAddress,
      decimals: apiBoost.earnedTokenDecimals,
      symbol: apiBoost.earnedToken,
      buyUrl: null,
      description: null,
      website: null,
      type: 'erc20',
    };
    temporaryWrappedtokenFix(token);
    sliceState.byChainId[chainId].byId[token.id] = token;
    sliceState.byChainId[chainId].interestingBalanceTokenIds.push(token.id);
  }
}

function addVaultToState(
  state: BeefyState,
  sliceState: WritableDraft<TokensState>,
  chain: ChainEntity,
  vault: VaultConfig
) {
  const chainId = chain.id;
  if (sliceState.byChainId[chainId] === undefined) {
    sliceState.byChainId[chainId] = {
      byId: {},
      interestingBalanceTokenIds: [],
      native: null,
      wnative: null,
    };
  }

  const oracleToken = getOracleTokenFromLegacyVaultConfig(selectChainById(state, chainId), vault);
  if (sliceState.byChainId[chainId].byId[oracleToken.id] === undefined) {
    sliceState.byChainId[chainId].byId[oracleToken.id] = oracleToken;
    sliceState.byChainId[chainId].interestingBalanceTokenIds.push(oracleToken.id);
  }

  // add earned token data
  const earnedTokenId = vault.earnedToken;
  if (sliceState.byChainId[chainId].byId[earnedTokenId] === undefined) {
    // gov vaults yield native tokens
    if (vault.isGovVault) {
      // all gov vaults yield Wnative tokens

      const token: TokenEntity =
        vault.earnedToken === sliceState.byChainId[chainId].native
          ? {
              id: earnedTokenId,
              chainId: chainId,
              address: null,
              decimals: chain.walletSettings.nativeCurrency.decimals,
              symbol: vault.earnedToken,
              type: 'native',
              buyUrl: sliceState.byChainId[chainId].byId[earnedTokenId]?.buyUrl ?? null,
              website: sliceState.byChainId[chainId].byId[earnedTokenId]?.website ?? null,
              description: sliceState.byChainId[chainId].byId[earnedTokenId]?.description ?? null,
            }
          : {
              id: earnedTokenId,
              chainId: chainId,
              decimals: vault.earnedTokenDecimals ?? 18,
              contractAddress: vault.earnedTokenAddress,
              symbol: vault.earnedToken,
              type: 'erc20',
              buyUrl: sliceState.byChainId[chainId].byId[earnedTokenId]?.buyUrl ?? null,
              website: sliceState.byChainId[chainId].byId[earnedTokenId]?.website ?? null,
              description: sliceState.byChainId[chainId].byId[earnedTokenId]?.description ?? null,
            };
      temporaryWrappedtokenFix(token);
      sliceState.byChainId[chainId].byId[token.id] = token;
      sliceState.byChainId[chainId].interestingBalanceTokenIds.push(token.id);
    } else {
      const token: TokenEntity = {
        id: earnedTokenId,
        chainId: chainId,
        contractAddress: vault.earnedTokenAddress,
        decimals: 18, // TODO: not sure about that
        symbol: vault.earnedToken,
        buyUrl: sliceState.byChainId[chainId].byId[earnedTokenId]?.buyUrl ?? null,
        website: sliceState.byChainId[chainId].byId[earnedTokenId]?.website ?? null,
        description: sliceState.byChainId[chainId].byId[earnedTokenId]?.description ?? null,
        type: 'erc20',
      };
      // temporaryWrappedtokenFix(token);
      sliceState.byChainId[chainId].byId[token.id] = token;
      sliceState.byChainId[chainId].interestingBalanceTokenIds.push(token.id);
    }
  }
}

const wrappedTokenFixes = [
  {
    chainId: 'fantom',
    tokenId: 'WFTM',
    contractAddress: '0x21be370d5312f44cb42ce377bc9b8a0cef1a4c83',
  },
  {
    chainId: 'polygon',
    tokenId: 'WMATIC',
    contractAddress: '0x0d500b1d8e8ef31e21c99d1db9a6444d3adf1270',
  },
];

/**
 * Some Wrapped tokens do not have a contractAddress set
 * This is a temporary fix
 */
function temporaryWrappedtokenFix(token: TokenEntity) {
  // only concern erc20 assets
  if (!isTokenErc20(token)) {
    return;
  }
  // contract address has been set
  if (token.contractAddress) {
    return;
  }

  // not a wrapped token
  if (!token.id.startsWith('W')) {
    console.error(`Erc20 token without a contract address: ${token.id} (${token.chainId})`);
    return;
  }

  for (const fix of wrappedTokenFixes) {
    if (token.chainId === fix.chainId && token.id === fix.tokenId) {
      token.contractAddress = fix.contractAddress;
      console.warn(`Fixed ${token.id} (${token.chainId}) contract address: ${fix.contractAddress}`);
      return;
    }
  }
  console.error(`Unfixed wrapped token without a contract address: ${token.id} (${token.chainId})`);
}
