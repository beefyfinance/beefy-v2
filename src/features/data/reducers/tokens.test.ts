import { fetchAllBoosts, FulfilledAllBoostsPayload } from '../actions/boosts';
import { fetchAllPricesAction } from '../actions/prices';
import { fetchAddressBookAction } from '../actions/tokens';
import { fetchAllVaults, FulfilledAllVaultsPayload } from '../actions/vaults';
import { BeefyAPITokenPricesResponse } from '../apis/beefy';
import { selectChainNativeToken } from '../selectors/tokens';
import { getBeefyTestingStore } from '../utils/test-utils';
import { initialTokensState, tokensSlice } from './tokens';

describe('Tokens slice tests', () => {
  it('should update state on fulfilled vault list', async () => {
    const store = await getBeefyTestingStore();
    const state = store.getState();
    const payload: FulfilledAllVaultsPayload = {
      state,
      byChainId: {
        harmony: [
          // have one gov vault
          {
            id: 'one-bifi-gov',
            logo: 'single-assets/BIFI.png',
            name: 'BIFI Earnings Pool',
            token: 'BIFI',
            tokenDescription: 'Beefy.Finance',
            tokenAddress: '0x6ab6d61428fde76768d7b45d8bfeec19c6ef91a8',
            tokenDecimals: 18,
            tokenDescriptionUrl:
              'https://docs.beefy.finance/moo/ecosystem/bifi-token/tokenomics-and-governance',
            earnedToken: 'WONE',
            earnedTokenAddress: '0x5b96bbaca98d777cb736dd89a519015315e00d02',
            earnContractAddress: '0x5b96bbaca98d777cb736dd89a519015315e00d02',
            poolAddress: '0x5b96bbaca98d777cb736dd89a519015315e00d02',
            pricePerFullShare: 1,
            isGovVault: true,
            tvl: 0,
            oracle: 'tokens',
            oracleId: 'BIFI',
            oraclePrice: 0,
            status: 'active',
            platform: 'Beefy.Finance',
            assets: ['BIFI'],
            risks: [
              'COMPLEXITY_LOW',
              'BATTLE_TESTED',
              'IL_NONE',
              'MCAP_SMALL',
              'AUDIT',
              'CONTRACTS_VERIFIED',
            ],
            stratType: 'SingleStake',
            callFee: 0.5,
            withdrawalFee: '0%',
            createdAt: 1623706144,
            network: 'harmony',
          },

          // have one lp vault
          {
            id: 'sushi-one-1ygg-1eth',
            name: '1YGG-1ETH LP',
            token: '1YGG-1ETH LP',
            tokenDescription: 'Sushi',
            tokenAddress: '0x40112850EFd90e9e17b56de35d86BdFf9f4d07Fd',
            tokenDecimals: 18,
            tokenDescriptionUrl: '#',
            earnedToken: 'mooSushiYGG-ETH',
            earnedTokenAddress: '0x69F683aa79ab78B95A9823Ace8b9771Ad18fFC71',
            earnContractAddress: '0x69F683aa79ab78B95A9823Ace8b9771Ad18fFC71',
            pricePerFullShare: 1,
            tvl: 0,
            oracle: 'lps',
            oracleId: 'sushi-one-1ygg-1eth',
            oraclePrice: 0,
            status: 'active',
            platform: 'Sushi',
            assets: ['YGG', 'ETH'],
            callFee: 0.25,
            withdrawalFee: '0%',
            addLiquidityUrl:
              'https://app.sushi.com/add/0x6983D1E6DEf3690C4d616b13597A09e6193EA013/0x63cf309500d8be0B9fDB8F1fb66C821236c0438c',
            buyTokenUrl:
              'https://app.sushi.com/swap?inputCurrency=0x6983D1E6DEf3690C4d616b13597A09e6193EA013&outputCurrency=0x63cf309500d8be0B9fDB8F1fb66C821236c0438c',
            stratType: 'StratLP',
            network: 'harmony',
            risks: [
              'COMPLEXITY_LOW',
              'BATTLE_TESTED',
              'IL_LOW',
              'MCAP_MEDIUM',
              'AUDIT',
              'CONTRACTS_VERIFIED',
            ],
          },
        ],
      },
    };
    const action = { type: fetchAllVaults.fulfilled, payload: payload };
    const newState = tokensSlice.reducer(initialTokensState, action);
    expect(newState).toMatchSnapshot();
  });

  it('should update state on fulfilled boosts list', () => {
    const payload: FulfilledAllBoostsPayload = {
      harmony: [
        // one with a partnership
        {
          id: 'moo_banana-pera',
          poolId: 'banana-banana-eol',
          name: 'Pera',
          logo: 'degens/BANANA.svg',
          earnedToken: 'PERA',
          earnedTokenDecimals: 18,
          earnedTokenAddress: '0xb9D8592E16A9c1a3AE6021CDDb324EaC1Cbc70d6',
          earnContractAddress: '0x38247fCE28480A7BEF2CB7aD134ce091Bd2E1a82',
          earnedOracle: 'tokens',
          earnedOracleId: 'PERA',
          partnership: true,
          status: 'closed',
          isMooStaked: true,
          partners: [
            {
              logo: 'stake/pera/logo.png',
              background: 'stake/pera/bg.png',
              text: "Pera Finance is a decentralized trading competition platform with multi-layered yield farming opportunities. Traders, liquidity providers and holders yield farm together through the DeFi's first decentralized trading competition. Every on-chain PERA transaction (transfer, trade, or liquidity addition/ removal) generates a 2 % transaction fee.",
              website: 'https://pera.finance/',
              social: {
                telegram: 'https://t.me/perafinance',
                twitter: 'https://twitter.com/perafinance',
              },
            },
          ],
        },

        // one closed boost
        {
          id: 'moo_aave-btc',
          poolId: 'aave-wbtc',
          name: 'Iron / Garuda / Fanatics',
          logo: 'single-assets/BTCB.svg',
          earnedToken: 'mooPolygonBIFI',
          earnedTokenDecimals: 18,
          earnedTokenAddress: '0xfEcf784F48125ccb7d8855cdda7C5ED6b5024Cb3',
          earnContractAddress: '0x20948Cad130c3D7B24d27CC66163b4aaed4684F0',
          earnedOracle: 'tokens',
          earnedOracleId: 'BIFI',
          partnership: true,
          status: 'closed',
          isMooStaked: true,
          partners: [],
        },
      ],
    };
    const action = { type: fetchAllBoosts.fulfilled, payload: payload };
    const state = tokensSlice.reducer(initialTokensState, action);
    expect(state).toMatchSnapshot();
  });

  it('should update state on fulfilled prices list', () => {
    const payload: BeefyAPITokenPricesResponse = {
      '1AAVE': 238.16568740268866,
      '1AXS': 79.04891966147113,
      '1BTC': 609115.1079533856,
      '1inch-1inch-bnb': 0,
      '2omb-2omb-ftm': 16.710317081168196,
      '2omb-2share-ftm': 171.84420463276956,
      // when rpc are down, the API might send us nulls
      // in this case, we say 1 token = $1
      'solarbeam-avax-wmovr': null,
      'solarbeam-bnb-busd': null,
    };
    const action = { type: fetchAllPricesAction.fulfilled, payload: payload };
    const state = tokensSlice.reducer(initialTokensState, action);
    expect(state).toMatchSnapshot();
  });

  it('should correctly handle mooFuse config', () => {
    const payload: FulfilledAllBoostsPayload = {
      fuse: [
        {
          id: 'moo_fusefi-wfuse-usdc-fuse',
          poolId: 'voltage-wfuse-usdc',
          name: 'Fuse',
          assets: ['USDC', 'FUSE'],
          earnedToken: 'mooFuse',
          earnedTokenDecimals: 18,
          earnedTokenAddress: '0x2C43DBef81ABa6b95799FD2aEc738Cd721ba77f3',
          earnContractAddress: '0x405EE7F4f067604b787346bC22ACb66b06b15A4B',
          earnedOracle: 'tokens',
          earnedOracleId: 'WFUSE',
          partnership: true,
          status: 'active',
          isMooStaked: true,
          partners: [
            {
              logo: 'stake/fuse/logo.png',
              background: 'stake/fuse/bg.png',
              text: 'Fuse is a platform featuring a fast and low-cost, Ethereum-compatible blockchain, a robust plug-and-play mobile-centric crypto payments infrastructure for creating token-based applications and a rapidly growing ecosystem of payments, decentralized finance and NFT projects.',
              website: 'https://www.fuse.io',
              social: {
                telegram: 'https://t.me/fuseio',
                twitter: 'https://twitter.com/Fuse_network',
              },
            },
          ],
        },
      ],
    };
    const action = { type: fetchAllBoosts.fulfilled, payload: payload };
    const state = tokensSlice.reducer(initialTokensState, action);
    expect(state).toMatchSnapshot();
  });

  it('should handle fuse-fuse config correctly', async () => {
    const store = await getBeefyTestingStore();
    const payload: FulfilledAllVaultsPayload = {
      state: store.getState(),
      byChainId: {
        fuse: [
          {
            id: 'fuse-fuse',
            logo: 'single-assets/FUSE.svg',
            name: 'FUSE',
            token: 'FUSE',
            tokenDescription: 'Beefy Delegator',
            tokenDecimals: 18,
            tokenDescriptionUrl: '#',
            earnedToken: 'mooFuse',
            earnedTokenAddress: '0x2C43DBef81ABa6b95799FD2aEc738Cd721ba77f3',
            earnContractAddress: '0x2C43DBef81ABa6b95799FD2aEc738Cd721ba77f3',
            pricePerFullShare: 1,
            tvl: 0,
            oracle: 'tokens',
            oracleId: 'WFUSE',
            oraclePrice: 0,
            depositsPaused: false,
            status: 'active',
            platform: 'Fuse',
            assets: ['WFUSE'],
            risks: [],
            stratType: 'SingleStake',
            withdrawalFee: '0%',
            buyTokenUrl:
              'https://app.fuse.fi/#/swap?inputCurrency=FUSE&outputCurrency=0xa722c13135930332Eb3d749B2F0906559D2C5b99',
            network: 'fuse',
            createdAt: 1641908745,
          },
        ],
      },
    };
    const action = { type: fetchAllVaults.fulfilled, payload: payload };
    const state = tokensSlice.reducer(initialTokensState, action);
    expect(state).toMatchSnapshot();
  });

  it('should update addressable native token addresses when fetching the address book', async () => {
    const store = await getBeefyTestingStore();
    await store.dispatch(fetchAddressBookAction({ chainId: 'metis' }));
    await store.dispatch(fetchAddressBookAction({ chainId: 'celo' }));
    await store.dispatch(fetchAddressBookAction({ chainId: 'bsc' }));
    const state = store.getState();
    const metisNative = selectChainNativeToken(state, 'metis');
    const celoNative = selectChainNativeToken(state, 'celo');
    const bscNative = selectChainNativeToken(state, 'bsc');
    expect(metisNative.address).toBe('0xDeadDeAddeAddEAddeadDEaDDEAdDeaDDeAD0000');
    expect(celoNative.address).toBe('0x471EcE3750Da237f93B8E339c536989b8978a438');
    expect(bscNative.address).toBe(null);
  });
});
