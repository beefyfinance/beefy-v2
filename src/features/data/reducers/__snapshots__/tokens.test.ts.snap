// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokens slice tests should correctly handle mooFuse config 1`] = `
Object {
  "byChainId": Object {
    "fuse": Object {
      "byId": Object {
        "mooFuse": Object {
          "buyUrl": null,
          "chainId": "fuse",
          "contractAddress": "0x2C43DBef81ABa6b95799FD2aEc738Cd721ba77f3",
          "decimals": 18,
          "description": null,
          "id": "mooFuse",
          "symbol": "mooFuse",
          "type": "erc20",
          "website": null,
        },
      },
      "interestingBalanceTokenIds": Array [
        "mooFuse",
      ],
      "native": null,
      "wnative": null,
    },
  },
  "prices": Object {
    "byTokenId": Object {},
  },
}
`;

exports[`Tokens slice tests should handle fuse-fuse config correctly 1`] = `
Object {
  "byChainId": Object {
    "fuse": Object {
      "byId": Object {
        "FUSE": Object {
          "address": null,
          "buyUrl": null,
          "chainId": "fuse",
          "decimals": 18,
          "description": null,
          "id": "FUSE",
          "symbol": "FUSE",
          "type": "native",
          "website": null,
        },
        "mooFuse": Object {
          "buyUrl": null,
          "chainId": "fuse",
          "contractAddress": "0x2C43DBef81ABa6b95799FD2aEc738Cd721ba77f3",
          "decimals": 18,
          "description": null,
          "id": "mooFuse",
          "symbol": "mooFuse",
          "type": "erc20",
          "website": null,
        },
      },
      "interestingBalanceTokenIds": Array [
        "FUSE",
        "mooFuse",
      ],
      "native": null,
      "wnative": null,
    },
  },
  "prices": Object {
    "byTokenId": Object {},
  },
}
`;

exports[`Tokens slice tests should update state on fulfilled boosts list 1`] = `
Object {
  "byChainId": Object {
    "harmony": Object {
      "byId": Object {
        "PERA": Object {
          "buyUrl": null,
          "chainId": "harmony",
          "contractAddress": "0xb9D8592E16A9c1a3AE6021CDDb324EaC1Cbc70d6",
          "decimals": 18,
          "description": null,
          "id": "PERA",
          "symbol": "PERA",
          "type": "erc20",
          "website": null,
        },
        "mooPolygonBIFI": Object {
          "buyUrl": null,
          "chainId": "harmony",
          "contractAddress": "0xfEcf784F48125ccb7d8855cdda7C5ED6b5024Cb3",
          "decimals": 18,
          "description": null,
          "id": "mooPolygonBIFI",
          "symbol": "mooPolygonBIFI",
          "type": "erc20",
          "website": null,
        },
      },
      "interestingBalanceTokenIds": Array [
        "PERA",
        "mooPolygonBIFI",
      ],
      "native": null,
      "wnative": null,
    },
  },
  "prices": Object {
    "byTokenId": Object {},
  },
}
`;

exports[`Tokens slice tests should update state on fulfilled prices list 1`] = `
Object {
  "byChainId": Object {},
  "prices": Object {
    "byTokenId": Object {
      "1AAVE": "238.16568740268866",
      "1AXS": "79.04891966147113",
      "1BTC": "609115.1079533856",
      "1inch-1inch-bnb": "0",
      "2omb-2omb-ftm": "16.710317081168196",
      "2omb-2share-ftm": "171.84420463276956",
      "solarbeam-avax-wmovr": "1",
      "solarbeam-bnb-busd": "1",
    },
  },
}
`;

exports[`Tokens slice tests should update state on fulfilled vault list 1`] = `
Object {
  "byChainId": Object {
    "harmony": Object {
      "byId": Object {
        "BIFI": Object {
          "buyUrl": null,
          "chainId": "harmony",
          "contractAddress": "0x6ab6d61428fde76768d7b45d8bfeec19c6ef91a8",
          "decimals": 18,
          "description": null,
          "id": "BIFI",
          "symbol": "BIFI",
          "type": "erc20",
          "website": null,
        },
        "WONE": Object {
          "buyUrl": null,
          "chainId": "harmony",
          "contractAddress": "0x5b96bbaca98d777cb736dd89a519015315e00d02",
          "decimals": 18,
          "description": null,
          "id": "WONE",
          "symbol": "WONE",
          "type": "erc20",
          "website": null,
        },
        "mooSushiYGG-ETH": Object {
          "buyUrl": null,
          "chainId": "harmony",
          "contractAddress": "0x69F683aa79ab78B95A9823Ace8b9771Ad18fFC71",
          "decimals": 18,
          "description": null,
          "id": "mooSushiYGG-ETH",
          "symbol": "mooSushiYGG-ETH",
          "type": "erc20",
          "website": null,
        },
        "sushi-one-1ygg-1eth": Object {
          "buyUrl": null,
          "chainId": "harmony",
          "contractAddress": "0x40112850EFd90e9e17b56de35d86BdFf9f4d07Fd",
          "decimals": 18,
          "description": null,
          "id": "sushi-one-1ygg-1eth",
          "symbol": "1YGG-1ETH LP",
          "type": "erc20",
          "website": null,
        },
      },
      "interestingBalanceTokenIds": Array [
        "BIFI",
        "WONE",
        "sushi-one-1ygg-1eth",
        "mooSushiYGG-ETH",
      ],
      "native": null,
      "wnative": null,
    },
  },
  "prices": Object {
    "byTokenId": Object {},
  },
}
`;
